File: packages/feed/packages/examples/feed-langgraph-agent/benchmark_runner.py
"""
Benchmark Runner for Feed LangGraph Agent

Runs this agent through benchmark simulations to measure performance.
Uses the same LangGraph logic but with a simulated A2A interface.
"""

import json
import logging
import sys
import os
from datetime import datetime
from typing import Dict, Any, List
import asyncio

# Setup logging
logging.basicConfig(
    level=logging.INFO,
    format='[%(asctime)s] [%(levelname)s] %(message)s',
    datefmt='%Y-%m-%d %H:%M:%S'
)
logger = logging.getLogger(__name__)

class BenchmarkSimulationA2A:
    """
    Simulation A2A client that replays pre-recorded benchmark data.
    Compatible with the agent's A2A interface expectations.
    """
    
    def __init__(self, snapshot: Dict[str, Any]):
        self.snapshot = snapshot
        self.current_tick = 0
        self.actions_taken = []
        
    def get_current_state(self):
        """Get game state for current tick"""
        if self.current_tick == 0:
            return self.snapshot['initialState']
        
        if self.current_tick - 1 < len(self.snapshot['ticks']):
            return self.snapshot['ticks'][self.current_tick - 1]['state']
        
        return self.snapshot['initialState']
    
    def advance_tick(self):
        """Move to next tick"""
        self.current_tick += 1
    
    async def get_predictions(self) -> Dict[str, Any]:
        """Get active prediction markets"""
        state = self.get_current_state()
        predictions = state.get('predictionMarkets', [])
        
        return {
            'predictions': [
                {
                    'id': m['id'],
                    'question': m['question'],
                    'yesPrice': m['yesPrice'],
                    'noPrice': m['noPrice'],
                    'yesShares': m['yesShares'],
                    'noShares': m['noShares'],
                    'liquidity': m.get('liquidity', 1000),
                    'totalVolume': m.get('totalVolume', 0),
                }
                for m in predictions
                if not m.get('resolved', False)
            ]
        }
    
    async def get_perpetuals(self) -> Dict[str, Any]:
        """Get perpetual markets"""
        state = self.get_current_state()
        perpetuals = state.get('perpetualMarkets', [])
        
        return {
            'perpetuals': [
                {
                    'ticker': m['ticker'],
                    'price': m['price'],
                    'priceChange24h': m.get('priceChange24h', 0),
                    'volume24h': m.get('volume24h', 0),
                    'openInterest': m.get('openInterest', 0),
                    'fundingRate': m.get('fundingRate', 0),
                }
                for m in perpetuals
            ]
        }
    
    async def get_feed(self, limit: int = 10) -> Dict[str, Any]:
        """Get social feed"""
        state = self.get_current_state()
        posts = state.get('posts', [])
        
        return {
            'posts': posts[-limit:] if len(posts) > limit else posts
        }
    
    async def get_portfolio(self) -> Dict[str, Any]:
        """Get agent portfolio (simplified)"""
        return {
            'balance': 10000,
            'positions': [],
            'pnl': 0
        }
    
    async def buy_shares(self, market_id: str, outcome: str, amount: float) -> Dict[str, Any]:
        """Buy prediction market shares"""
        action = {
            'type': 'buy_prediction',
            'tick': self.current_tick,
            'data': {
                'marketId': market_id,
                'outcome': outcome,
                'amount': amount
            }
        }
        self.actions_taken.append(action)
        
        return {
            'success': True,
            'shares': amount * 0.5,  # Simplified
            'positionId': f'pos-{len(self.actions_taken)}'
        }
    
    async def open_position(self, ticker: str, side: str, size: float, leverage: int) -> Dict[str, Any]:
        """Open perpetual position"""
        action = {
            'type': 'open_perp',
            'tick': self.current_tick,
            'data': {
                'ticker': ticker,
                'side': side,
                'size': size,
                'leverage': leverage
            }
        }
        self.actions_taken.append(action)
        
        return {
            'success': True,
            'positionId': f'perp-{len(self.actions_taken)}'
        }
    
    async def create_post(self, content: str, market_id: str = None) -> Dict[str, Any]:
        """Create social post"""
        action = {
            'type': 'create_post',
            'tick': self.current_tick,
            'data': {
                'content': content,
                'marketId': market_id
            }
        }
        self.actions_taken.append(action)
        
        return {
            'success': True,
            'postId': f'post-{len(self.actions_taken)}'
        }


async def run_benchmark(
    benchmark_file: str,
    output_dir: str,
    agent_module: str = 'agent'
) -> Dict[str, Any]:
    """
    Run agent through benchmark simulation
    
    Args:
        benchmark_file: Path to benchmark JSON file
        output_dir: Directory to save results
        agent_module: Python module name with agent (default: 'agent')
    
    Returns:
        Simulation results dictionary
    """
    
    logger.info('🎯 Starting LangGraph Agent Benchmark')
    logger.info(f'Benchmark: {benchmark_file}')
    logger.info(f'Output: {output_dir}')
    
    # 1. Load benchmark
    logger.info('📊 Loading benchmark data...')
    with open(benchmark_file, 'r') as f:
        snapshot = json.load(f)
    
    total_ticks = len(snapshot['ticks'])
    logger.info(f'  Loaded: {total_ticks} ticks')
    
    # 2. Create simulation A2A client
    logger.info('🔌 Creating simulation A2A client...')
    a2a_client = BenchmarkSimulationA2A(snapshot)
    
    # 3. Import agent
    logger.info(f'🤖 Loading agent from {agent_module}...')
    try:
        agent = __import__(agent_module)
        logger.info('  Agent loaded successfully')
    except ImportError as e:
        logger.error(f'Failed to import agent: {e}')
        raise
    
    # 4. Run simulation loop
    logger.info('🚀 Starting simulation loop...')
    tick_count = 0
    
    while a2a_client.current_tick < total_ticks:
        tick_count += 1
        
        if tick_count % 10 == 0 or tick_count == 1:
            progress = (tick_count / total_ticks) * 100
            logger.info(f'  Tick {tick_count}/{total_ticks} ({progress:.0f}%)')
        
        try:
            # Gather context
            predictions = await a2a_client.get_predictions()
            perpetuals = await a2a_client.get_perpetuals()
            feed = await a2a_client.get_feed(10)
            portfolio = await a2a_client.get_portfolio()
            
            # Build context for agent
            context = {
                'predictions': predictions.get('predictions', []),
                'perpetuals': perpetuals.get('perpetuals', []),
                'feed': feed.get('posts', []),
                'portfolio': portfolio
            }
            
            # Make decision using agent's logic
            # (This would call your agent's decision function)
            # For now, simplified example:
            await make_agent_decision(agent, context, a2a_client)
            
            # Advance tick
            a2a_client.advance_tick()
            
        except Exception as e:
            logger.warning(f'  ⚠️  Error on tick {tick_count}: {e}')
            # Continue anyway
            a2a_client.advance_tick()
    
    logger.info(f'✅ Simulation complete: {tick_count} ticks processed')
    
    # 5. Calculate metrics
    logger.info('📊 Calculating metrics...')
    metrics = calculate_metrics(a2a_client.actions_taken, snapshot)
    
    # 6. Save results
    logger.info('💾 Saving results...')
    os.makedirs(output_dir, exist_ok=True)
    
    result = {
        'id': f'langgraph-{datetime.now().isoformat()}',
        'agentType': 'langgraph',
        'benchmarkId': snapshot['id'],
        'ticksProcessed': tick_count,
        'actions': a2a_client.actions_taken,
        'metrics': metrics,
        'startTime': snapshot.get('createdAt', 0),
        'endTime': datetime.now().timestamp() * 1000
    }
    
    with open(os.path.join(output_dir, 'result.json'), 'w') as f:
        json.dump(result, f, indent=2)
    
    with open(os.path.join(output_dir, 'metrics.json'), 'w') as f:
        json.dump(metrics, f, indent=2)
    
    # Log summary
    logger.info('✅ Benchmark Complete!')
    logger.info('')
    logger.info('Results:')
    logger.info(f"  Total P&L: ${metrics.get('totalPnl', 0):.2f}")
    logger.info(f"  Prediction Accuracy: {metrics.get('predictionAccuracy', 0)*100:.1f}%")
    logger.info(f"  Perp Win Rate: {metrics.get('perpWinRate', 0)*100:.1f}%")
    logger.info(f"  Actions Taken: {len(a2a_client.actions_taken)}")
    logger.info('')
    logger.info(f"View results: {output_dir}")
    
    return result


async def make_agent_decision(agent, context: Dict[str, Any], a2a_client: BenchmarkSimulationA2A):
    """
    Make decision using agent's logic
    
    This is a simplified example - you would integrate your actual
    LangGraph agent decision-making here.
    """
    
    # Simple heuristic for demo:
    # - Look at predictions
    # - If any are mispriced, take action
    
    predictions = context.get('predictions', [])
    
    for market in predictions:
        yes_price = market.get('yesPrice', 0.5)
        
        # Simple strategy: buy YES if price < 0.4, buy NO if price > 0.6
        if yes_price < 0.4:
            await a2a_client.buy_shares(
                market_id=market['id'],
                outcome='YES',
                amount=100
            )
            break
        elif yes_price > 0.6:
            await a2a_client.buy_shares(
                market_id=market['id'],
                outcome='NO',
                amount=100
            )
            break


def calculate_metrics(actions: List[Dict[str, Any]], snapshot: Dict[str, Any]) -> Dict[str, Any]:
    """Calculate performance metrics from actions"""
    
    # Count action types
    prediction_actions = [a for a in actions if a['type'] == 'buy_prediction']
    perp_actions = [a for a in actions if a['type'] == 'open_perp']
    post_actions = [a for a in actions if a['type'] == 'create_post']
    
    # Simplified metrics calculation
    # In a real implementation, you would:
    # 1. Track position outcomes against ground truth
    # 2. Calculate actual P&L
    # 3. Measure timing of actions
    
    return {
        'totalPnl': 0,  # Would calculate from positions
        'predictionAccuracy': 0,  # Would compare against ground truth
        'perpWinRate': 0,  # Would track perp outcomes
        'actionsBreakdown': {
            'predictions': len(prediction_actions),
            'perpetuals': len(perp_actions),
            'posts': len(post_actions),
            'total': len(actions)
        },
        'timing': {
            'avgResponseTime': 0,
            'totalDuration': 0
        }
    }


async def run_multiple(
    benchmark_file: str,
    output_dir: str,
    runs: int,
    agent_module: str = 'agent'
):
    """Run multiple benchmark iterations"""
    
    logger.info(f'🔄 Running {runs} benchmark iterations')
    
    results = []
    
    for i in range(runs):
        logger.info('')
        logger.info('━' * 50)
        logger.info(f'Run {i+1}/{runs}')
        logger.info('━' * 50)
        
        run_dir = os.path.join(output_dir, f'run-{i+1}')
        result = await run_benchmark(benchmark_file, run_dir, agent_module)
        results.append(result)
        
        # Delay between runs
        await asyncio.sleep(1)
    
    # Calculate comparison
    avg_pnl = sum(r['metrics'].get('totalPnl', 0) for r in results) / runs
    avg_accuracy = sum(r['metrics'].get('predictionAccuracy', 0) for r in results) / runs
    
    comparison = {
        'runs': results,
        'comparison': {
            'avgPnl': avg_pnl,
            'avgAccuracy': avg_accuracy,
            'totalRuns': runs
        }
    }
    
    with open(os.path.join(output_dir, 'comparison.json'), 'w') as f:
        json.dump(comparison, f, indent=2)
    
    logger.info('')
    logger.info('━' * 50)
    logger.info('🏆 ALL BENCHMARKS COMPLETE')
    logger.info('━' * 50)
    logger.info(f'  Runs: {runs}')
    logger.info(f'  Avg P&L: ${avg_pnl:.2f}')
    logger.info(f'  Avg Accuracy: {avg_accuracy*100:.1f}%')
    logger.info('')
    logger.info(f'Results: {output_dir}')


def main():
    """CLI entry point"""
    
    import argparse
    
    parser = argparse.ArgumentParser(description='Run LangGraph agent through benchmark')
    parser.add_argument('--benchmark', required=True, help='Path to benchmark JSON file')
    parser.add_argument('--output', default=f'./benchmark-results/{int(datetime.now().timestamp())}',
                       help='Output directory for results')
    parser.add_argument('--runs', type=int, default=1, help='Number of runs (default: 1)')
    parser.add_argument('--agent', default='agent', help='Agent module name (default: agent)')
    
    args = parser.parse_args()
    
    try:
        if args.runs == 1:
            asyncio.run(run_benchmark(args.benchmark, args.output, args.agent))
        else:
            asyncio.run(run_multiple(args.benchmark, args.output, args.runs, args.agent))
    except Exception as e:
        logger.error(f'❌ Benchmark failed: {e}')
        import traceback
        traceback.print_exc()
        sys.exit(1)


if __name__ == '__main__':
    main()

File: packages/feed/packages/examples/feed-typescript-agent/src/benchmark-runner.ts
/**
 * Benchmark Runner for Autonomous Feed Agent
 *
 * Runs this agent through benchmark simulations to measure performance.
 * Uses the same decision-making logic but with a simulated A2A interface.
 */

import dotenv from "dotenv";

dotenv.config({ path: ".env.local" });

import fs from "node:fs";
import path from "node:path";

// Benchmark types (defined locally for examples)
// NOTE: Benchmark runner is disabled until simulation modules are available as packages
interface BenchmarkGameSnapshot {
  markets: unknown[];
  posts: unknown[];
  timestamp: number;
  ticks: Array<{ tick: number; markets: unknown[]; posts: unknown[] }>;
}

interface SimulationResult {
  id: string;
  totalTrades: number;
  totalPnL: number;
  winRate: number;
  avgROI: number;
  metrics: {
    totalPnl: number;
    predictionMetrics: {
      accuracy: number;
    };
    optimalityScore: number;
  };
}

const LOG_DIR = "./logs";
const LOG_FILE = path.join(LOG_DIR, "benchmark.log");

function log(message: string, level: "info" | "warn" | "error" = "info") {
  const timestamp = new Date().toISOString();
  const logLine = `[${timestamp}] [${level.toUpperCase()}] ${message}\n`;

  console.log(logLine.trim());

  if (!fs.existsSync(LOG_DIR)) {
    fs.mkdirSync(LOG_DIR, { recursive: true });
  }
  fs.appendFileSync(LOG_FILE, logLine);
}

async function loadBenchmark(
  benchmarkFile: string,
): Promise<BenchmarkGameSnapshot> {
  const data = fs.readFileSync(benchmarkFile, "utf-8");
  return JSON.parse(data);
}

async function runBenchmark(
  benchmarkFile: string,
  outputDir: string,
): Promise<SimulationResult> {
  log("🎯 Starting Autonomous Agent Benchmark");
  log(`Benchmark: ${benchmarkFile}`);
  log(`Output: ${outputDir}`);

  // 1. Load benchmark
  log("📊 Loading benchmark data...");
  const snapshot = await loadBenchmark(benchmarkFile);
  log(`  Loaded: ${snapshot.ticks.length} ticks`);

  // 2. Dynamic import of benchmark modules
  log("🔧 Loading simulation modules...");

  // Benchmark simulation modules are not available as packages.
  // They need to be migrated to @feed/testing or similar before this runner
  // can execute locally.
  throw new Error(
    "Benchmark runner requires simulation modules that are not yet available as packages. " +
      "Please use the main app benchmark runner instead.",
  );

  // Placeholder types for when modules are available
  // const { SimulationEngine } = await import('@feed/testing/benchmark/SimulationEngine');
  // const { SimulationA2AInterface } = await import('@feed/testing/benchmark/SimulationA2AInterface');
  // const { MetricsVisualizer } = await import('@feed/testing/benchmark/MetricsVisualizer');

  // NOTE: All code below is unreachable due to throw above, but kept for reference
  // when simulation modules become available as packages
  /*
  // 3. Create simulation engine
  log('🎮 Creating simulation engine...');
  const simConfig = {
    snapshot,
    agentId: 'feed-typescript-agent',
    fastForward: true,
    responseTimeout: 30000,
  };

  const engine = new SimulationEngine(simConfig);

  // 4. Create A2A interface
  log('🔌 Creating simulation A2A interface...');
  const a2aInterface = new SimulationA2AInterface(
    engine,
    'feed-typescript-agent'
  );

  // 5. Initialize agent components
  log('🧠 Initializing agent memory and decision maker...');
  const memory = new AgentMemory({ maxEntries: 20 });

  const strategy = (process.env.AGENT_STRATEGY ?? 'balanced') as
    | 'conservative'
    | 'balanced'
    | 'aggressive'
    | 'social';
  const decisionMaker = new AgentDecisionMaker({
    strategy,
    groqApiKey: process.env.GROQ_API_KEY,
    anthropicApiKey: process.env.ANTHROPIC_API_KEY,
    openaiApiKey: process.env.OPENAI_API_KEY,
  });

  log(`  Strategy: ${strategy}`);
  log(`  LLM Provider: ${decisionMaker.getProvider()}`);

  // 6. Run simulation loop
  log('🚀 Starting simulation loop...');
  let tickCount = 0;
  const totalTicks = snapshot.ticks.length;

  while (engine.getGameState().tick < totalTicks) {
    tickCount++;

    if (tickCount % 10 === 0 || tickCount === 1) {
      log(
        `  Tick ${tickCount}/${totalTicks} (${((tickCount / totalTicks) * 100).toFixed(0)}%)`
      );
    }

    await (async () => {
      // Gather context via A2A interface
      const portfolioResponse = (await a2aInterface.sendRequest(
        'a2a.getPortfolio'
      )) as { balance: number; positions: A2APerpPosition[]; pnl: number };
      const portfolio: {
        balance: number;
        positions: A2APerpPosition[];
        pnl: number;
      } = {
        balance: portfolioResponse?.balance ?? 10000,
        positions: portfolioResponse?.positions ?? [],
        pnl: portfolioResponse?.pnl ?? 0,
      };

      const predictionsResponse = (await a2aInterface.sendRequest(
        'a2a.getPredictions'
      )) as { predictions: PredictionMarket[] };
      const perpetualsResponse = (await a2aInterface.sendRequest(
        'a2a.getPerpetuals'
      )) as { perpetuals: PerpMarket[] };
      const feedResponse = (await a2aInterface.sendRequest('a2a.getFeed', {
        limit: 10,
      })) as { posts: FeedPost[] };

      const markets = {
        predictions: predictionsResponse?.predictions ?? [],
        perps: perpetualsResponse?.perpetuals ?? [],
      };

      const recentMemory = memory.getRecent(5);

      // Make decision
      const decision = await decisionMaker.decide({
        portfolio,
        markets,
        feed: feedResponse ? { posts: feedResponse.posts } : { posts: [] },
        memory: recentMemory,
      });

      // Execute action (if not HOLD)
      if (decision.action !== 'HOLD') {
        // Create a wrapper that matches A2AActionClient interface
        const actionClient: A2AActionClient = {
          buyShares: async (
            marketId: string,
            outcome: 'YES' | 'NO',
            amount: number
          ) => {
            const result = await a2aInterface.buyShares(
              marketId,
              outcome,
              amount
            );
            return result as Record<string, JsonValue>;
          },
          sellShares: async (positionId: string, shares: number) => {
            const result = await a2aInterface.sellShares(positionId, shares);
            return result as Record<string, JsonValue>;
          },
          openPosition: async (
            ticker: string,
            side: 'LONG' | 'SHORT',
            amount: number,
            leverage: number
          ) => {
            // Convert LONG/SHORT to long/short for SimulationA2AInterface
            const simSide = side === 'LONG' ? 'long' : 'short';
            const result = await a2aInterface.openPosition(
              ticker,
              simSide,
              amount,
              leverage
            );
            return result as Record<string, JsonValue>;
          },
          closePosition: async (positionId: string) => {
            const result = await a2aInterface.closePosition(positionId);
            return result as Record<string, JsonValue>;
          },
          createPost: async (content: string, type?: string) => {
            const result = await a2aInterface.createPost(content, type);
            return result as Record<string, JsonValue>;
          },
          createComment: async (postId: string, content: string) => {
            const result = await a2aInterface.createComment(postId, content);
            return result as Record<string, JsonValue>;
          },
        };

        const result = await executeAction(actionClient, decision);

        if (result.success) {
          memory.add({
            action: decision.action,
            params: decision.params ?? {},
            result: result.data ?? {},
            timestamp: Date.now(),
          });
        }
      }

      // Advance tick
      engine.advanceTick();
    })().catch((error: Error) => {
      log(`  ⚠️  Error on tick ${tickCount}: ${error.message}`, 'warn');
      // Continue anyway
      engine.advanceTick();
    });
  }

  log(`✅ Simulation complete: ${tickCount} ticks processed`);

  // 7. Get final results
  log('📊 Collecting results...');
  const result = await engine.run();

  // 8. Save results
  log('💾 Saving results...');
  if (!fs.existsSync(outputDir)) {
    fs.mkdirSync(outputDir, { recursive: true });
  }

  fs.writeFileSync(
    path.join(outputDir, 'result.json'),
    JSON.stringify(result, null, 2)
  );

  fs.writeFileSync(
    path.join(outputDir, 'metrics.json'),
    JSON.stringify(result.metrics, null, 2)
  );

  // 9. Generate visualizations
  log('📈 Generating visualizations...');
  await MetricsVisualizer.visualizeSingleRun(result, {
    outputDir,
    generateHtml: true,
    generateCsv: true,
    generateCharts: false,
  });

  log('✅ Benchmark Complete!');
  log('');
  log('Results:');
  log(`  Total P&L: $${result.metrics.totalPnl.toFixed(2)}`);
  log(
    `  Prediction Accuracy: ${(result.metrics.predictionMetrics.accuracy * 100).toFixed(1)}%`
  );
  log(
    `  Perp Win Rate: ${(result.metrics.perpMetrics.winRate * 100).toFixed(1)}%`
  );
  log(`  Optimality Score: ${result.metrics.optimalityScore.toFixed(1)}%`);
  log(
    `  Duration: ${(result.metrics.timing.totalDuration / 1000).toFixed(1)}s`
  );
  log('');
  log(`View report: file://${path.join(outputDir, 'index.html')}`);

  return result;
  */
}

async function runMultiple(
  _benchmarkFile: string,
  _outputDir: string,
  _runs: number,
): Promise<{ runs: SimulationResult[]; comparison: unknown }> {
  // Benchmark runner is disabled - see runBenchmark function
  throw new Error(
    "Benchmark runner requires simulation modules that are not yet available as packages.",
  );

  // Unreachable code kept for reference
  /*
  log(`🔄 Running ${runs} benchmark iterations`);

  const results: SimulationResult[] = [];

  for (let i = 0; i < runs; i++) {
    log('');
    log('━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━');
    log(`Run ${i + 1}/${runs}`);
    log('━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━');

    const runDir = path.join(outputDir, `run-${i + 1}`);
    const result = await runBenchmark(benchmarkFile, runDir);
    results.push(result);

    // Delay between runs
    await new Promise((resolve) => setTimeout(resolve, 1000));
  }

  // Calculate comparison metrics
  const calculateAverage = (getValue: (r: SimulationResult) => number) =>
    results.reduce((sum, r) => sum + getValue(r), 0) / runs;

  const avgPnl = calculateAverage((r) => r.metrics.totalPnl);
  const avgAccuracy = calculateAverage(
    (r) => r.metrics.predictionMetrics.accuracy
  );
  const avgOptimality = calculateAverage((r) => r.metrics.optimalityScore);

  const bestRun = results.reduce((best, r) =>
    r.metrics.totalPnl > best.metrics.totalPnl ? r : best
  );
  const worstRun = results.reduce((worst, r) =>
    r.metrics.totalPnl < worst.metrics.totalPnl ? r : worst
  );

  const comparison = {
    runs: results,
    comparison: {
      avgPnl,
      avgAccuracy,
      avgOptimality,
      bestRun: bestRun.id,
      worstRun: worstRun.id,
    },
  };

  fs.writeFileSync(
    path.join(outputDir, 'comparison.json'),
    JSON.stringify(comparison, null, 2)
  );

  log('');
  log('━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━');
  log('🏆 ALL BENCHMARKS COMPLETE');
  log('━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━');
  log(`  Runs: ${runs}`);
  log(`  Avg P&L: $${avgPnl.toFixed(2)}`);
  log(`  Avg Accuracy: ${(avgAccuracy * 100).toFixed(1)}%`);
  log(`  Avg Optimality: ${avgOptimality.toFixed(1)}%`);
  log('');
  log(`Results: ${outputDir}`);
  */
}

// CLI
async function main() {
  const args = process.argv.slice(2);

  const benchmarkFile = args
    .find((a) => a.startsWith("--benchmark="))
    ?.split("=")[1];
  const outputDir =
    args.find((a) => a.startsWith("--output="))?.split("=")[1] ??
    `./benchmark-results/${Date.now()}`;
  const runs = Number.parseInt(
    args.find((a) => a.startsWith("--runs="))?.split("=")[1] ?? "1",
    10,
  );

  if (!benchmarkFile) {
    console.error("Error: --benchmark is required");
    console.log("\nUsage:");
    console.log(
      "  bun run src/benchmark-runner.ts --benchmark=path/to/benchmark.json [options]",
    );
    console.log("\nOptions:");
    console.log("  --output=path/to/results  Output directory");
    console.log("  --runs=5                  Number of runs (default: 1)");
    process.exit(1);
  }

  if (runs === 1) {
    await runBenchmark(benchmarkFile, outputDir);
  } else {
    await runMultiple(benchmarkFile, outputDir, runs);
  }
}

main();
